server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -145,10 +145,29 @@ public RecoveryResponse recoverToTarget() throws IOException {
             }
             assert targetShardRouting.initializing() : "expected recovery target to be initializing but was " + targetShardRouting;
         }, shardId + " validating recovery target ["+ request.targetAllocationId() + "] registered ", shard, cancellableThreads, logger);
-
-        try (Closeable ignored = shard.acquireTranslogRetentionLock()) {
+        /*
+         * DISCUSS:
```

Comment:
wait, if you acquire the snapshot before you check the `isSequenceNumberBasedRecovery` lucene will not trim away anything since you prevented the policy from going forward. It has point in time semantics so I don't understand what you mean by stuff get's trimmed. Merged that are in-flight will have seen previous retention limits so that's fine. Pending merges will pick up their retention limits right before they execute which is also fine? I think they only issue I can see is that `setCheckpointOfSafeCommit` accepts a smaller value than the number of ops we keep around and then things go sideways. Otherwise we should be just fine? I must be missing something

Reply:
@s1monw, I should have explained this better.

The problem is that the method `isTranslogReadyForSequenceNumberBasedRecovery` may see operations that are not protected by the retention policy, then decide to go with seq# based recovery. However, an actual snapshot for sending ops might not have those "unprotected" ops if the searcher is refreshed to the newly merged segment. I have a test to illustrate this scenario.

```
    public void testPeerRecoveryRetentionLock() throws Exception {
        AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
        try (Store store = createStore();
             InternalEngine engine = createEngine(config(defaultSettings, store, createTempDir(), newMergePolicy(), null, null, globalCheckpoint::get))) {
            MapperService mapper = createMapperService("test");

            engine.index(indexForDoc(createParsedDoc("1", null)));
            engine.delete(new Engine.Delete("test", "1", newUid("1"), primaryTerm.get()));
            engine.flush(); // The first segment

            engine.index(indexForDoc(createParsedDoc("2", null)));
            engine.flush(); // The second segment

            // Schedules a merge that claims the first segment.
            // This merge is scheduled before the retention lock and finished in between snapshot1 and snapshot2
            engine.blockMerges();
            globalCheckpoint.set(engine.getLocalCheckpointTracker().getCheckpoint());
            engine.syncTranslog(); // Advances the safe commit

            Thread mergeThread = new Thread(() -> {
                try {
                    engine.forceMerge(false, 1, true, false, false);
                } catch (IOException e) {
                    throw new AssertionError(e);
                }
            });
            mergeThread.start();

            Closeable retentionLock = engine.acquireRetentionLockForPeerRecovery();
            Engine.IndexCommitRef safeCommit = engine.acquireSafeIndexCommit();

            // Assume the replica starts from seq# 0, isSequenceNumberBasedRecovery is true because the snapshot has seq# 0,1,2
            try (Translog.Snapshot snapshotToDecide = engine.newLuceneChangesSnapshot("test", mapper, 0, 2, false)) {
                assertThat(snapshotToDecide.next().seqNo(), equalTo(0L));
                assertThat(snapshotToDecide.next().seqNo(), equalTo(1L));
                assertThat(snapshotToDecide.next().seqNo(), equalTo(2L));
            }
            engine.unblockMerges();
            mergeThread.join();
            assertBusy(() -> assertThat(engine.getMergeStats().getCurrent(), equalTo(0L)));

            // This searcher opens a newly merged segment which does not have seq#0 and seq#1
            engine.refresh("test");
            try (Translog.Snapshot snapshotToSendOps = engine.newLuceneChangesSnapshot("test", mapper, 0, 2, false)) {
                assertThat(snapshotToSendOps.next().seqNo(), equalTo(0L)); // This fails
                assertThat(snapshotToSendOps.next().seqNo(), equalTo(1L));
                assertThat(snapshotToSendOps.next().seqNo(), equalTo(2L));
            }
            IOUtils.close(retentionLock, safeCommit);
        }
    }

```

Reply:
I agree with Simon that we should take different approach then the one we had with the translog, now that we reason about trimming soft deletes in terms seq# (the translog uses generation size + time). My suggestion is to use the following logic:
1) Acquire a retention lock.
2) Use the current min retained op from the retention policy to check if the request op range is above it. If so, we do a ops based recovery, if not we don't (even though, in theory we can get lucky and have some operations in lucene before the min retained op). 

I also think we should not pre-flight the requested range like `isTranslogReadyForSequenceNumberBasedRecovery` does. If you the numbers indicate that an ops based recovery is possible, we go for it. We should then verify we managed to bring the target shard to where we need to be (local checkpoint >= the end of the ops range we send) and if not we fail the recovery. IMO we can start to rely on retention semantics and simplify the logic. 

Note that this brings another interesting discussion: should we even support translog based ops based recovery? We should talk about it but I tend to say that we should but given it a whole different code path than the lucene based one. We will then use the soft delete setting to decide which path to follow and nothing else. Specifically - no fall back from lucene ops to translog ops. It's one or another.

Reply:
Yeah, I made this change as you suggested. However, we can use the min retained ops directly as the number of retained operations can be changed - it's a dynamic setting. I used the max_seq_no that has been exposed to the merge policy as the baseline. This value will be baked into the commit and bootstrap when opening an engine.